### ISSUE TITLE
Cosmos: Errors should be a concrete custom type

### ISSUE THREAD

**Issue by thomastaylor312** [Azure/azure-sdk-for-rust#128]
Right now all of the cosmos client methods return a `Box<dyn Error>`. This leads to [less than ideal error handling](https://github.com/deislabs/bindle-server-azure/blob/fa4ddd63e4d457fcf6bf622afb04f4aea41fe962/src/storage/mod.rs#L355). Ideally it should be a custom type that at least allows you to easily access the HTTP status code.

This may overlap with some of the discussions that have been had around error handling

**Comment by MindFlavor** [comment]
I really like [your code](https://github.com/deislabs/bindle-server-azure/blob/fa4ddd63e4d457fcf6bf622afb04f4aea41fe962/src/storage/mod.rs#L355). I think you have written the code this  library should have 👍. 

We should return meaningful, in context error responses (like you do returning `AlreadyExists`, `ItemNotFound` etc...). We should preserve the HTTP codes for the user to inspect should they like to but they should not be forced to map an HTTP code to the actual explanation.

We could start with an "inner" error like this (thinking out loud):

```rust
struct AzureError {
   HTTPRequest request,
   HTTPResponse response,
   Context context, // this should allow the user to know what he was doing, like "Inserting a document on XXX collection"
   string Description
}
```

And then having the various crates to build on in:

```rust
enum InsertEntityFatalError {
   Duplicate(AzureError),
   PrerequisiteFailed(AzureError), // for etag and such
   // and so on...
}

// this is needed to express transient errors
enum InsertEntityTransientError {
   TransientError(AzureError),
   //   and so on...
}

// and wrap them up
enum InsertEntityError {
   InsertEntityFatalError(InsertEntityFatalError),
   InsertEntityTransientError(InsertEntityTransientError)
}
```

Ideally we should generalize a way to express a "transient" error in the whole library (so the user know they need to retry). Something like:

```rust
trait TransientError {}

impl TransientError for InsertEntityTransientError {}
```

What do you think?

**Comment by thomastaylor312** [comment]
I think it may be simpler to consume if we don't have nested enums as you'd end up writing nested `match` blocks to introspect the error. I don't know all the different error cases for Cosmos, so maybe that enum would be a bit too big, but I would represent the errors in your example as:

```rust
struct CosmosError {
  kind: ErrorKind,
  // Any additional human readable info
  message: Option<String>,
  status_code: http::status::StatusCode,
}

enum ErrorKind {
  EntityExists,
  // Contains the prereq that failed
  PrerequisiteFailed(String),
  // When the error was caused by something not related to Cosmos
  AzureError(AzureError),
  Throttled(ThrottleDetails),
  // and so on...
}

impl CosmosError {
  fn new(kind: ErrorKind, message: Option<String>) -> Self {}
  fn is_transient(&self) -> bool {
    match self.kind {
      ErrorKind::AzureError(_) | ErrorKind::Throttled(_) => true,
      _ => false,
    }
}
```

As you can see in the example, I don't think every variant needs an internal `AzureError`. The reason being is that some types of errors (like `Duplicate`) don't really need the extra embedded error. I am not entirely sure we need something to indicate if it is transient or not as each user could handle each type of error differently, but I included it anyway

**Comment by rylev** [comment]
Not all errors are going to be HTTP related. The user might supply an invalid parameter (which we detect before sending to the API), or perhaps the JSON they receive back is not what they're expecting and we cannot parse it for them (i.e., the request itself was correct but the response was simply different than what the user thought it would be).. 

I worry we'll end up tying our error too tightly to HTTP.

**Comment by thomastaylor312** [comment]
That's fine too! It seemed to me that currently the main errors come from http status codes. But if #129 gets fixed, then that will change. If that is the case, I'd embed the underlying http error in the enum variants where it is needed

**Comment by cataggar** [comment]
Fixed by #625
